virtcontainers/utils/utils_linux_ppc64le.go

```diff
@@ -0,0 +1,95 @@
+// +build ppc64le
+
+// Copyright (c) 2018 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package utils
+
+import (
+	"crypto/rand"
+	"fmt"
+	"math/big"
+	"os"
+	"syscall"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+// from <linux/vhost.h>
+// VHOST_VSOCK_SET_GUEST_CID = _IOW(VHOST_VIRTIO, 0x60, __u64)
+
+// _IOC_WRITE is 1 for arch generic and 4 for powerpc
+// Reference: https://github.com/torvalds/linux/blob/6f0d349d922ba44e4348a17a78ea51b7135965b1/arch/powerpc/include/uapi/asm/ioctl.h
+const ioctlVhostVsockSetGuestCid = 0x8008AF60
+
+var ioctlFunc = Ioctl
+
+// maxUInt represents the maximum valid value for the context ID.
+// The upper 32 bits of the CID are reserved and zeroed.
+// See http://stefanha.github.io/virtio/
+var maxUInt uint64 = 1<<32 - 1
+
+func Ioctl(fd uintptr, request, data uintptr) error {
+	if _, _, errno := unix.Syscall(unix.SYS_IOCTL, fd, request, data); errno != 0 {
+		//uintptr(request)
+		//uintptr(unsafe.Pointer(&arg1)),
+		//); errno != 0 {
+		return os.NewSyscallError("ioctl", fmt.Errorf("%d", int(errno)))
+	}
+
+	return nil
+}
+
+// FindContextID finds a unique context ID by generating a random number between 3 and max unsigned int (maxUint).
+// Using the ioctl VHOST_VSOCK_SET_GUEST_CID, findContextID asks to the kernel if the given
+// context ID (N) is available, when the context ID is not available, incrementing by 1 findContextID
+// iterates from N to maxUint until an available context ID is found, otherwise decrementing by 1
+// findContextID iterates from N to 3 until an available context ID is found, this is the last chance
+// to find a context ID available.
+// On success vhost file and a context ID greater or equal than 3 are returned, otherwise 0 and an error are returned.
+// vhost file can be used to send vhost file decriptor to QEMU. It's the caller's responsibility to
+// close vhost file descriptor.
+//
+// Benefits of using random context IDs:
+// - Reduce the probability of a *DoS attack*, since other processes don't know whatis the initial context ID
+//   used by findContextID to find a context ID available
+//
+func FindContextID() (*os.File, uint64, error) {
```

Comment:
This function is the same across all architectures so it would be good to find a way to avoid this arch-specific copy.

Reply:
@jodh-intel: Updated my PR. Ptal ...